feat(engine): Impelemented Convex Hull constructor

TadaTeruki opened the pull request:
# Overview
- added convex hull constructor as a new action

## How I tested

test workflow and data (another branch):
https://github.com/reearth/reearth-flow/blob/tada-action-process-with-testcase/engine/runtime/action-processor/src/geometry/testcase/convex-hull.yml

## Screenshot

![Screenshot from 2025-02-05 20-20-55](https://github.com/user-attachments/assets/97bc3a4d-fe3a-4175-8545-3d70e1e62f16)

## Memo

各Convex Hullの属性は、group byの対象となるもののみ継承しました

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced convex hull processing capabilities that enable users to generate convex hulls from geometric data with configurable grouping options.
	- Expanded schema support with built-in processors available in multiple language configurations for enhanced geometric processing.
- **Tests**
	- Added a new workflow configuration and accompanying command-line utility to facilitate end-to-end testing of the convex hull processing pipeline.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+    buffer_2d: HashMap<String, Vec<Feature>>,
+}
+
+fn get_attributes(feature: &Feature, attributes: &[Attribute]) -> Vec<AttributeValue> {
```
Please add it here.
https://github.com/reearth/reearth-flow/blob/main/engine/runtime/types/src/feature.rs#L259

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+        let feature = &ctx.feature;
+        let geometry = &feature.geometry;
+        if geometry.is_empty() {
+            fw.send(ctx.new_with_feature_and_port(ctx.feature.clone(), DEFAULT_PORT.clone()));
```
```suggestion
            fw.send(ctx.new_with_feature_and_port(ctx.feature.clone(), REJECTED_PORT.clone()));
```

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+#[derive(Debug, Clone)]
+pub struct ConvexHullConstructor {
+    group_by: Option<Vec<Attribute>>,
+    buffer_2d: HashMap<String, Vec<Feature>>,
```
```suggestion
    buffer_2d: HashMap<AttributeValue, Vec<Feature>>,
```

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+                        .collect::<Vec<_>>()
+                        .join("\t")
+                } else {
+                    "_all".to_string()
```
```suggestion
     AttributeValue::String("_all".to_string())
```

miseyu replied:
```suggestion
                    AttributeValue::String("_all".to_string())
```

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+            } else {
+                continue;
+            };
+
```
We want to initialize the metadata and id, so be sure to do Feature::new.

miseyu commented:
@TadaTeruki 
Please check back when you have time!

miseyu commented:
Please E2E test cases should be placed here!
https://github.com/reearth/reearth-flow/tree/tada-action-process-with-testcase/engine/runtime/tests

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+
+impl ProcessorFactory for ConvexHullConstructorFactory {
+    fn name(&self) -> &str {
+        "ConvexHullConstructor"
```
```suggestion
        "ConvexHullAccumulator"
```
https://docs.safe.com/fme/html/FME-Form-Documentation/FME-Transformers/Transformers/hullaccumulator.htm

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+#[derive(Debug, Clone, Default)]
+pub struct ConvexHullConstructorFactory;
+
+impl ProcessorFactory for ConvexHullConstructorFactory {
```
```suggestion
impl ProcessorFactory for ConvexHullAccumulator Factory {
```

miseyu commented on engine/runtime/action-processor/src/geometry/convex_hull_constructor.rs:
```diff
@@ -0,0 +1,209 @@
+}
+
+#[derive(Debug, Clone)]
+pub struct ConvexHullConstructor {
```
```suggestion
pub struct ConvexHullAccumulator {
```

TadaTeruki commented:
- The E2E test is now just a shell script replacement for the test command and is not automated.
```
$ bash runtime/tests/testcase/geometry/convex_hull_accumulator.sh <output_path>
```

- The `get_attributes` function only performed `feature.attributes.get(attr)` for multiple attribute keys. Since it was redundant,  I removed it and replaced it with method chaining instead of adding it as a method of `Feature`.

miseyu reviewed (approved):
LGTM
